### PR TITLE
Ypersky -  pvc creation deletion performance

### PR DESCRIPTION
**Pull request by ypersky1980**
Current PR impements the following : 

1) PVC creation and PVC deletion tests were merged into 1 flow : 
         create PVC with samples  , take measurements 
         deleted PVC with samples , take measurements 
         analyze the measurements ( calculate standard deviation) 
         push results to codespeed dashboard 

2) An improvement was made - all the PVC now have data written to them after the creation and prior to deletion. 

3) test_multiple_pvc_deletion_measurement_performance  moved from test_pvc_multiple_creation file to test_pvc_creation_deletion_performance and was  enhanced with writing data to PVCs as well.  
    So in the miltiple deletion case we will also delete PVCs with data and not empty PVCs.

### CODE REVIEW

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        if deletion_st_deviation_percent > accepted_deletion_deviation_percent:
+            raise ex.PerformanceException(
+                f"{msg_prefix} PVC deletion time deviation is {deletion_st_deviation_percent}%"
+                f"and is greater than the allowed {accepted_deletion_deviation_percent}%."
+            )
+
+        log.info(
+            f"{msg_prefix} The standard deviation percent for deletion of {len(deletion_time_measures)} sampled PVCs "
+            f"is {deletion_st_deviation_percent}%."
+        )
```
Comment (Avilir) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:167]: instead of duplicate the code for creation and deletion, you can create a function which accept (values, title), the title is create/delete.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+            pod_objs.append(pod_obj)
+
+        # Get pvc_name, require pvc_name to fetch deletion time data from log
+        threads = list()
+        for pvc_obj in pvc_objs:
+            process = threading.Thread(target=pvc_obj.reload)
+            process.start()
+            threads.append(process)
+        for process in threads:
+            process.join()
```
Comment (Avilir) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:250]: are you sure running 120 threads simultaneously will be faster ?

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        threads = list()
+        for pvc_obj in pvc_objs:
+            process1 = threading.Thread(target=pvc_name_list.append(pvc_obj.name))
+            process2 = threading.Thread(target=pv_name_list.append(pvc_obj.backed_pv))
+            process1.start()
+            process2.start()
+            threads.append(process1)
+            threads.append(process2)
+        for process in threads:
+            process.join()
```
Comment (Avilir) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:262]: same as above, just that here you are running 240 threads - does the client (Jenkins slave) can handle that ?

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        msg_prefix = f"Interface: {self.interface}, PVC size: {pvc_size}."
+
+        log.info(f"{msg_prefix} Start creating new 120 PVCs")
+
+        pvc_objs = helpers.create_multiple_pvcs(
+            sc_name=self.sc_obj.name,
+            namespace=defaults.ROOK_CLUSTER_NAMESPACE,
+            number_of_pvc=number_of_pvcs,
+            size=pvc_size,
+        )
```
Comment (Avilir) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:225]: you are creating the PVC's one by one and not in bulk, for bulk creation you need to add the argument `burst=true` (or just `true` at the end).

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        for process in threads:
+            process.join()
+        log.info(f"{msg_prefix} Preparing to delete 120 PVC")
+
+        # Delete PVC
+        for pvc_obj, pod_obj in zip(pvc_objs, pod_objs):
+            pod_obj.delete(wait=True)
+            pvc_obj.delete()
+            pvc_obj.ocp.wait_for_delete(pvc_obj.name)
+
```
Comment (Avilir) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:270]: this deletion is one by one, is this what you intend to do ?

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 2 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:250]: @Avilir  I don't know, but this is how it worked before. So I thought that the essense of this test is to create and delete in threads. I did not want to change that, only added writing data before the deletion.

File: ocs_ci/helpers/performance_lib.py [hunk 6]
```diff
@@ -86,3 +89,111 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
...
+    command = f"oc --kubeconfig {kubeconfig} -n {namespace} {cmd}"
+    return run_command(command, out_format="list")
+
+
+def get_log_names(interface):
+    """
+    Finds names for log files pods in which logs for clone creation are located
+    For CephFS: 2 pods that start with "csi-cephfsplugin-provisioner" prefix
+    For RBD: 2 pods that start with "csi-rbdplugin-provisioner" prefix
+
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:126]: @Avilir Done, thank you!

File: ocs_ci/helpers/performance_lib.py [hunk 7]
```diff
@@ -86,3 +89,111 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
...
+        if provisioning_name in line:
+            log_names.append(line.split()[0])
+
+    logger.info(f"The logs pods are : {log_names}")
+    return log_names
+
+
+def measure_pvc_creation_time(interface, pvc_name, start_time):
+    """
+    Args:
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:149]: @Avilir Done.

File: ocs_ci/helpers/performance_lib.py [hunk 8]
```diff
@@ -86,3 +89,111 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
         output = output.split("\n")  # convert output to list
         output.pop()  # remove last empty element from the list
     return output
+
+
+def run_oc_command(cmd, namespace):
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:94]: @Avilir Thank you, added that.

File: ocs_ci/helpers/performance_lib.py [hunk 9]
```diff
@@ -86,3 +89,111 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
...
+
+    for line in pods:
+        if provisioning_name in line:
+            log_names.append(line.split()[0])
+
+    logger.info(f"The logs pods are : {log_names}")
+    return log_names
+
+
+def measure_pvc_creation_time(interface, pvc_name, start_time):
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:147]: @Avilir  The difference between my function in performance_lib.py and the ocs_ci.helpers.helpers.measure_pvc_creation_time() is that we are not going over all the log file but starting to search from start time parameter. 
Also their function does not take into consideration the fact that somoetimes there are multiple start time and multiple end time records in the log. 
I am also usiung this function in the multiple clones test, in which the time of the test is critical. 

Therefore I need this function as it is written in the performance_lib to make sure that 
1) we are not going over all the log file to search for the records 
2) We calculate the creation time as the difference between the last "end time" appearance and the first "start time" appearance. And the old helpers function does not implement that. 

I did not want to change the function in ocs_ci.helpers , since there might be other tests depending on it. 
After this PR is merged, I will also take those functions from multi clone test ( in order not to duplicate code). 
I'll do it in a separate PR since this PR is already big enough.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 1 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:167]: @Avilir  I'll try to implement such a function.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 4 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:225]: @Avilir I've changed it, will run now to test it.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 5 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:270]: @Avilir I think so, since this is what we were measuring before my changes. 
Then, I suppose, we should also create them one by one, without burst=true paramters passed in helpers.create_multiple_pvcs  . 
So I thihk that I'll remove that burst=True , right ?

File: ocs_ci/helpers/performance_lib.py [hunk 10]
```diff
@@ -86,3 +89,111 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
         output = output.split("\n")  # convert output to list
         output.pop()  # remove last empty element from the list
     return output
+
+
+def run_oc_command(cmd, namespace):
+    """
+    Running an 'oc' command
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:96]: @keesturam I've added more detailed description.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 5 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:270]: After our coversation it has been decided to implement now bulk creation ( with burst = True) and serial deletion. 
An option of bulk deletion will be implemented in the future.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 1 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:167]: @Avilir  Done. 
def process_time_measurements( self, action_name, time_measures, accepted_deviation_percent, msg_prefix ) was implemented.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+from ocs_ci.helpers import helpers, performance_lib
+from ocs_ci.ocs import defaults, constants
+from ocs_ci.utility.performance_dashboard import push_to_pvc_time_dashboard
+
+
+log = logging.getLogger(__name__)
+
+
+@performance
+class TestPVCCreationDeletionPerformance(E2ETest):
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:23]: @keesturam after talking to Avi, change to PASTest.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+            pv_name = pvc_obj.backed_pv
+            pvc_reclaim_policy = pvc_obj.reclaim_policy
+
+            pod_obj = self.write_file_on_pvc(pvc_obj)
+            pod_obj.delete(wait=True)
+            teardown_factory(pvc_obj)
+            logging.info(f"{msg_prefix} Start deleting PVC number {i + 1}")
+            pvc_obj.delete()
+            pvc_obj.ocp.wait_for_delete(pvc_obj.name)
+            if pvc_reclaim_policy == constants.RECLAIM_POLICY_DELETE:
```
[the remark this answers is not in the document]
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:97]: @keesturam 
I've added the following condition to log print : 

logging.info(
                    f"Reclaim policy of PVC {pvc_obj.name} is not Delete;"
                    f" therefore not measuring deletion time for this PVC."
                )

**Comment by ypersky1980** [comment]
Currently it's not possible to run PR validation job since openstack.psi.redhat.com is down. 
The log file of the successfull run ( 8 out of 8 tests passed) is available here: 10.70.39.233:/ypersky_pr/ypersky_44_test_pvc_creation_deletion_perf_2

**Review by Avilir — APPROVED** [review]
Nice work @ypersky1980

File: ocs_ci/helpers/performance_lib.py [hunk 13]
```diff
@@ -86,3 +89,123 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
         output = output.split("\n")  # convert output to list
         output.pop()  # remove last empty element from the list
     return output
+
+
+def run_oc_command(cmd, namespace=constants.OPENSHIFT_STORAGE_NAMESPACE):
+    """
+    Running an 'oc' command
+    This function is needed in Performance tests in order to be able to run a ceparate command within the test
```
Comment (keesturam) [inline ocs_ci/helpers/performance_lib.py:97]: replace ceperate with separate

File: ocs_ci/helpers/performance_lib.py [hunk 14]
```diff
@@ -86,3 +89,123 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
...
+    elif os.path.exists(cluster_dir_kubeconfig):
+        kubeconfig = cluster_dir_kubeconfig
+    else:
+        kubeconfig = None
+
+    command = f"oc --kubeconfig {kubeconfig} -n {namespace} {cmd}"
+    return run_command(command, out_format="list")
+
+
+def get_log_names(interface):
```
Comment (keesturam) [inline ocs_ci/helpers/performance_lib.py:123]: `get_logfile_names` would improve readability

File: ocs_ci/helpers/performance_lib.py [hunk 15]
```diff
@@ -86,3 +89,123 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
...
+    """
+
+    Measure PVC creation time, provided pvc name and time after which the PVC was created
+
+    Args:
+        interface : an interface (RBD or CephFS) to run on
+        pvc_name : Name of the pvc for which we measure the time
+        start_time: Time from which and on to search the relevant logs
+
+    Returns:
```
Comment (keesturam) [inline ocs_ci/helpers/performance_lib.py:165]: Please update the type of arguments and returns. For eg., int for return

File: ocs_ci/helpers/performance_lib.py [hunk 16]
```diff
@@ -86,3 +89,123 @@ def run_command(cmd, timeout=600, out_format="string", **kwargs):
...
+                f"logs {l} -c csi-provisioner --since-time={start_time}",
+                "openshift-storage",
+            )
+        )
+
+    format = "%H:%M:%S.%f"
+
+    st = None
+    et = None
+    # look for start time and end time of clone creation. The start/end line may appear in log several times
```
Comment (keesturam) [inline ocs_ci/helpers/performance_lib.py:183]: The comment mentions 'clone creation'. I believe this is a generic function. maybe it was copied from somewhere? if so, please update the comment to remove info related to clone.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+import datetime
+import pytest
+import ocs_ci.ocs.exceptions as ex
+import threading
+import statistics
+
+from ocs_ci.framework.testlib import performance
+from ocs_ci.ocs.perftests import PASTest
+
+from concurrent.futures import ThreadPoolExecutor
```
Comment (keesturam) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:14]: can you please move this to line 10. 
It is a good practice to sort importing general libraries at the beginning followed by ocs-ci libraries.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+
+        Args:
+            action_name (str) Name of the action for which these measurements were collected; used for the logging
+            time_measures: A list of time measurements
+            accepted_deviation_percent: Accepted deviation percent, if the standard  deviation of the provided time
+                measurements is bigger than this value, the test fails
+            msg_prefix: A string for comprehensive logging
+
+        Returns
+            The average value of the provided time measurements
```
Comment (keesturam) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:154]: Please add the type of value being returned

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        average = statistics.mean(time_measures)
+        log.info(
+            f"{msg_prefix} The average {action_name} time for the sampled {len(time_measures)} "
+            f"PVCs is {average} seconds."
+        )
+
+        st_deviation = statistics.stdev(time_measures)
+        st_deviation_percent = st_deviation / average * 100.0
```
Comment (keesturam) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:163]: I'm not sure why we do this - can't we simply say any st-deviation above 0.5 is not acceptable?

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+        # Get PVC deletion time
+        pvc_deletion_time = helpers.measure_pv_deletion_time_bulk(
+            interface=self.interface, pv_name_list=pv_name_list
+        )
+        log.info(
+            f"{msg_prefix} {number_of_pvcs} bulk deletion time is {pvc_deletion_time}"
+        )
+
+        # accepted deletion time is 2 secs for each PVC
+        accepted_pvc_deletion_time = number_of_pvcs * 2
```
Comment (keesturam) [inline tests/e2e/performance/test_pvc_creation_deletion_performance.py:285]: what is the bulk deletion time we are observing?
since we run deletion in parallel, deletion should take much lesser time. for eg., 120 PVC deletion in your case would be 240 seconds but I guess it should get completed in less than 120 seconds.

Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:285]: @keesturam  actually the deletion is serial in this case, not parralel - Avi and I had this conversation as well last week. 
The plan is to handle ( if it's needed) parallel deletion in one of the future PRs, since currently we don't have a function that deletes mltiple PVCs as bulk.

File: ocs_ci/helpers/performance_lib.py (hunk 13 above)
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:97]: @keesturam done, thank you.

File: ocs_ci/helpers/performance_lib.py (hunk 15 above)
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:165]: @keesturam done.

File: ocs_ci/helpers/performance_lib.py (hunk 14 above)
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:123]: @keesturam Done, indeed better. thanks!

File: ocs_ci/helpers/performance_lib.py (hunk 16 above)
Comment (ypersky1980) [reply ocs_ci/helpers/performance_lib.py:183]: @keesturam Done, thank you!

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 17 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:14]: @keesturam Done.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 18 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:154]: @keesturam Done.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 19 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:163]: @keesturam Here I think that even if the calculations here look a little bit "too much" - it's more comprehemsive to compare standard deviation percents. Is it ok to leave it like this ?

Comment (keesturam) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:162]: @ypersky1980 sure.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 20 above)
Comment (keesturam) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:284]: okay, maybe raise a github issue to track this requirement

**Review by keesturam — APPROVED** [review]

**Comment by keesturam** [comment]
Great work @ypersky1980 on implementing these changes.

File: tests/e2e/performance/test_pvc_creation_deletion_performance.py (hunk 20 above)
Comment (ypersky1980) [reply tests/e2e/performance/test_pvc_creation_deletion_performance.py:284]: https://github.com/red-hat-storage/ocs-ci/issues/4097   - I've opened a git hub issue and put a "performance" label there.

**Comment by ypersky1980** [comment]
The test run successfully after the latest commits, 8/8 passed, the log for the run is available here:  

10.70.39.233:/ypersky_pr/ypersky_44_test_pvc_creation_deletion_perf_1

**Review by Avilir — APPROVED** [review]
